db/dbScript.sql

```diff
@@ -0,0 +1,393 @@
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp"; --it is needed for uuid generator (uuid_generate_v4()) to be available
+
+--Basically, we dynamically create enums if they do not exist
+BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'provider_type') THEN
+        CREATE TYPE provider_type AS ENUM ('local', 'google');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'account_type') THEN
+        CREATE TYPE account_type AS ENUM ('user', 'gym', 'admin');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'own_status') THEN
+        CREATE TYPE own_status AS ENUM ('pending', 'approved', 'rejected');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'rec_type') THEN
+        CREATE TYPE rec_type AS ENUM ('main', 'alternative');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'not_type') THEN
+        CREATE TYPE not_type AS ENUM ('message', 'alert', 'reminder');
+    END IF;
+END $$;
+
+--Since we will have a lot of gyms stored (potentially) it does not make sense to store working hours for each of them as a text
+--Especially since there are significantly smaller potential patterns [(opening hours, closing hours) pairs] in comparison to the potential number of gyms stored
+CREATE TABLE if not exists Working_hours
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),--uuid is analogous to id, but uuid is pseudo-random and the length and format of uuids are the same
+  open_from TIME NOT NULL,
+  open_until TIME NOT NULL,
+  PRIMARY KEY (id),
+  UNIQUE (open_from),
+  UNIQUE (open_until),
+  check (open_until > open_from)
+);
+
+CREATE TABLE if not exists Currency
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),
+  name VARCHAR(10) NOT NULL,
+  code CHAR(3) NOT NULL,
+  PRIMARY KEY (id),
+  UNIQUE (name),
+  UNIQUE (code)
+);
+
+CREATE TABLE if not exists Account
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),
+  username VARCHAR(40) NOT null check (length(username) > 5),
+  email VARCHAR(60) not NULL,
+  first_name VARCHAR(60) NOT NULL,
+  last_name VARCHAR(60) NOT NULL,
+  created_at TIMESTAMP NOT null check (created_at <= NOW()),
+  provider provider_type NOT NULL, --we want to differentiate between google accounts and local accounts
+  password_hash CHAR(60) NOT NULL, --we are storing hash of the password that we got from bycrypt
+  type account_type NOT NULL, --we want to differentate between user, admin and gym accounts
+  created_by uuid, --admin accounts can be created only by another admin accounts. in this case we want to register who created the admin account
+  PRIMARY KEY (id),
+  FOREIGN KEY (created_by) REFERENCES Account(id),
+  UNIQUE (username),
+  UNIQUE (email)
+);
+create index if not exists idx_account_username on Account(username);
+create index if not exists idx_account_email on Account(email);
+create index if not exists idx_account_type on Account("type");
+
+create table if not exists Country
```

Comment:
Why do we need a separate Country table? It might be more efficient to use an ENUM or a VARCHAR field directly if we're only storing the country name.

Reply:
I do not think that an Enum that contains names of all the countries is a good idea, and with the Varchar it will not be in the 3rd normal form (cause countries with the same name can be factored out)
Plus, say, if the app has never been used anywhere but in Croatia, it does not make sense to store all the countries in the db which would be the case with enums